mirrord/intproxy/src/lib.rs

```diff
@@ -171,52 +192,88 @@ impl IntProxy {
         !self.task_txs.layers.is_empty()
     }
 
-    /// Runs main event loop of this proxy.
+    /// Runs the main event loop till a failure or success happens, if the failure is manageable, it
+    /// goes in failover state starting to update every layer with the error content on every new or
+    /// pending task. In failover state it continues to accept connection from layers
     /// Expects to accept the first layer connection within the given `first_timeout`.
     /// Exits after `idle_timeout` when there are no more layer connections.
     pub async fn run(
-        mut self,
+        self,
+        first_timeout: Duration,
+        idle_timeout: Duration,
+    ) -> Result<(), ProxyStartupError> {
+        match self.run_inner(first_timeout, idle_timeout).await {
+            Ok(()) => Ok(()),
+            Err(ProxyFailure::Startup(error)) => Err(error),
+            Err(ProxyFailure::Runtime(failover_strategy)) => {
+                tracing::warn!(
+                    "Managed exception {}, proxy is entering in failover state...",
+                    failover_strategy.fail_cause()
+                );
+                failover_strategy.run(idle_timeout, idle_timeout).await
+            }
+        }
+    }
+
+    /// Runs main event loop of this proxy.
+    /// Expects to accept the first layer connection within the given `first_timeout`.
+    /// Exits after `idle_timeout` when there are no more layer connections.
+    async fn run_inner(
+        self,
         first_timeout: Duration,
         idle_timeout: Duration,
-    ) -> Result<(), IntProxyError> {
+    ) -> Result<(), ProxyFailure> {
         self.task_txs
             .agent
             .send(ClientMessage::SwitchProtocolVersion(
                 mirrord_protocol::VERSION.clone(),
             ))
             .await;
 
+        let mut proxy = self;
+
         loop {
             tokio::select! {
-                Some((task_id, task_update)) = self.background_tasks.next() => {
+                Some((task_id, task_update)) = proxy.background_tasks.next() => {
                     tracing::trace!(
                         %task_id,
                         ?task_update,
                         "Received a task update",
                     );
-
-                    self.handle_task_update(task_id, task_update).await?;
+                    let update_res = proxy.handle_task_update(task_id, task_update).await;
+                    tracing::trace!(%task_id, ?update_res, "task updated");
```

Comment:
```suggestion
                    let update_res = proxy.handle_task_update(task_id, task_update).await;
```
This would be a lot of spam :eyes: